src/main/java/spoon/MavenLauncher.java

```diff
@@ -108,656 +125,144 @@ public MavenLauncher(String mavenProject, String m2RepositoryPath, SOURCE_TYPE s
-			}
-			if (sourcePath == null) {
-				sourcePath = Paths.get(directory.getAbsolutePath(), "src", "test", "java").toString();
-			}
-			File source = new File(sourcePath);
-			if (source.exists()) {
-				output.add(source);
-			}
-			File generatedSource = Paths.get(directory.getAbsolutePath(), "target", "generated-test-sources").toFile();
-			if (generatedSource.exists()) {
-				output.add(generatedSource);
-			}
-			for (InheritanceModel module : modules) {
-				output.addAll(module.getTestDirectories());
-			}
-			return output;
-		}
-
-		/**
-		 * Extract the variable from a string
-		 */
-		private String extractVariable(String value) {
-			if (value != null && value.startsWith("$")) {
-				value = getProperty(value.substring(2, value.length() - 1));
-			}
-			return value;
-		}
-
-		private List<Version> getVersionsFromM2(String groupId, String artifactId) {
-			String groupIdPath = groupId.replace(".", "/");
-			File[] fileVersion = Paths.get(m2RepositoryPath, groupIdPath, artifactId).toFile().listFiles(file -> {
-				if (!file.isDirectory()) {
-					return false;
-				}
-				String version = file.getName();
-				return Paths.get(m2RepositoryPath, groupIdPath, artifactId, version, artifactId + "-" + version + ".jar").toFile().exists();
-			});
-			if (fileVersion == null) {
-				return Collections.emptyList();
-			}
-			List<Version> versions = Arrays.stream(fileVersion).map(f -> new Version(f.getName())).collect(Collectors.toList());
-			versions.sort(Comparator.reverseOrder());
-			return versions;
-		}
-
-		class Version implements Comparable<Version> {
-			String version;
-			int major = -1;
-			int minor = -1;
-			int incremental = -1;
-			int build = -1;
-			String qualifier;
-
-			Version(String version) {
-				this.version = version;
-				int buildIndex = version.indexOf("-");
-				if (buildIndex != -1) {
-					String build = version.substring(buildIndex + 1);
-					try {
-						this.build = Integer.parseInt(build);
-					} catch (NumberFormatException e) {
-						this.qualifier = build;
-					}
-					version = version.substring(0, buildIndex);
-				}
-				String[] splitVersion = version.split("\\.");
-				try {
-					this.major = Integer.parseInt(splitVersion[0]);
-				} catch (NumberFormatException ignore) {
-				}
-				if (splitVersion.length > 1) {
-					try {
-						this.minor = Integer.parseInt(splitVersion[1]);
-					} catch (NumberFormatException ignore) {
-					}
-				}
-				if (splitVersion.length > 2) {
+			} catch (Exception e) {
+				e.printStackTrace();
+				throw new SpoonException("Failed to read classpath written in temporary file " + classPathFile.getAbsolutePath() + ".");
+			} finally {
+				if (br != null) {
 					try {
-						this.incremental = Integer.parseInt(splitVersion[2]);
-					} catch (NumberFormatException ignore) {
+						br.close();
+					} catch (IOException e) {
+						e.printStackTrace(); //These Exceptions occur after the classpath has been read, so it should not prevent normal execution.
```

Comment:
Throw instead a SpoonException maybe?

Reply:
Yes. I'll do that.

Reply:
Actually you can use a try with resources to be sure it's closed at the end.
Something like:
```
try (br = new BufferedReader(new FileReader(classPathFile));) {
...
}
```

Reply:
And maybe add the throws IOException to the method itself if it's not public anymore. You'll manage those elsewhere.

Reply:
Ok this should be fixed now.